### PR TITLE
Behavior Tests Parent PR

### PR DESCRIPTION
**Pull request by sdreyer**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: client/src/featureform/resources.py [hunk 1]
```diff
@@ -657,11 +354,12 @@ def type(self) -> str:
         return "SPARK_OFFLINE"
 
     def serialize(self) -> bytes:
+        print(self.store_config)
```
Comment (sdreyer) [inline client/src/featureform/resources.py:357]: remove

File: tests/end_to_end/features/spark_e2e.feature [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    Examples: GCS
+      | filesize | filetype |  storage_provider |     bucket     |
+      |  small   |   csv    |        gcs        |featureform-test|
+      |  small   |  parquet |        gcs        |featureform-test|
+      |  large   |   csv    |        gcs        |featureform-test|
+      |  large   |  parquet |        gcs        |featureform-test|
+      |  small   | directory|        gcs        |featureform-test|
+      |  large   | directory|        gcs        |featureform-test|
+
+  Scenario Outline: Register a file with invalid stores
```
Comment (anthonylasso) [inline tests/end_to_end/features/spark_e2e.feature:43]: being able to test like this is awesome

File: tests/end_to_end/features/steps/client.py [hunk 3]
```diff
@@ -4,3 +4,26 @@
 @given("Featureform is installed")
```
Comment (anthonylasso) [inline tests/end_to_end/features/steps/client.py:4]: would it be possible to target a particular version?

File: tests/end_to_end/features/steps/spark_registeration.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    context.client.apply()
+
+
+@then("I should be able to pull the transformation as a dataframe")
+def step_impl(context):
+    df = context.client.dataframe(context.transformation)
+    assert (
+        len(df) == context.file_length
+    ), f"Expected {context.file_length} rows, got {len(df)} rows"
+    print(len(df))
```
Comment (anthonylasso) [inline tests/end_to_end/features/steps/spark_registeration.py:156]: do we want to leave these prints for debugging?

File: tests/end_to_end/features/steps/client.py (hunk 3 above)
Comment (sdreyer) [reply tests/end_to_end/features/steps/client.py:4]: We could, but this is mainly to test against whatever is on the users current branch

File: .github/workflows/behave-tests.yml [hunk 5]
```diff
@@ -29,14 +29,21 @@ jobs:
       - name: Install Protobuf
         run: sudo snap install protobuf --classic
 
+      - name: Build Docker Container
+        run: docker build . -t featureform
```
[the remark this answers is not in the document]
Comment (sdreyer) [reply .github/workflows/behave-tests.yml]: yeah or some other kubernetes instance

File: tests/end_to_end/features/provider_updates.feature [hunk 6]
```diff
@@ -0,0 +1,97 @@
+Feature: Provider Updates
+
+  Scenario Outline: Rotate S3 Credentials
+    Given Featureform is installed
+    When I create a "hosted" "insecure" client for "localhost:7878"
+    And I create the AWS credentials with an access key "<access_key>" and secret key "<secret_key>"
+    And I register S3 with name "<name>", bucket_name "<bucket_name>", bucket_region "<bucket_region>", path "<path>"
+    And I create the AWS credentials with an access key "<repl_access_key>" and secret key "<repl_secret_key>"
```
Comment (ahmadnazeri) [inline tests/end_to_end/features/provider_updates.feature:8]: Should this be done before the register S3?

File: tests/end_to_end/features/provider_updates.feature [hunk 7]
```diff
@@ -0,0 +1,97 @@
+Feature: Provider Updates
+
+  Scenario Outline: Rotate S3 Credentials
+    Given Featureform is installed
+    When I create a "hosted" "insecure" client for "localhost:7878"
+    And I create the AWS credentials with an access key "<access_key>" and secret key "<secret_key>"
+    And I register S3 with name "<name>", bucket_name "<bucket_name>", bucket_region "<bucket_region>", path "<path>"
```
Comment (ahmadnazeri) [inline tests/end_to_end/features/provider_updates.feature:7]: This is duplicated on line 9

File: tests/end_to_end/features/provider_updates.feature [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    Examples: Change Region
+      | name    | bucket_name | bucket_region | path    | access_key   | secret_key   | repl_name  | repl_bucket_path | repl_path   | repl_region  | repl_access_key   | repl_secret_key   | exp_bucket_name | exp_bucket_region | exp_path  | exp_access_key   |  exp_secret_key  |
+      | s3_1    |     bucket  |    us-east-1  | my/path |  access_key  |  secret_key  |   test     |   bucket         |   my/path   |   us-east-2  |    access_key     |    secret_key     |    bucket       |     us-east-1     |  my/path  |    access_key    |    secret_key    |
+
+
+  Scenario Outline: Azure Blob Store Registration
+    Given Featureform is installed
+    When I create a "hosted" "insecure" client for "localhost:7878"
+    And I register Blob Store with name "<name>", account_name "<account_name>", account_key "<account_key>", container_name "<container_name>", path "<path>"
+    And I register Blob Store with name "<name>", account_name "<repl_account_name>", account_key "<repl_account_key>", container_name "<repl_container_name>", path "<repl_path>"
```
Comment (ahmadnazeri) [inline tests/end_to_end/features/provider_updates.feature:42]: any reason we register the blob store twice?

File: tests/end_to_end/features/provider_updates.feature [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      | name  | account_name | account_key | container_name | path | repl_account_name | repl_account_key | repl_container_name | repl_path | exp_account_name | exp_account_key | exp_container_name | exp_path | exception |
+      | blob2 | account_name | account_key | container_name | path |  account_name     |  account_key     |  container_name     | path2      | account_name     | account_key     | container_name    | path     | None      |
+
+
+  Scenario Outline: Rotate GCS Credentials
+    Given Featureform is installed
+    When I create a "hosted" "insecure" client for "localhost:7878"
+    And I create the GCP credentials with a project id "<project_id>" and credentials path "<credentials_path>"
+    And I register GCS with name "<name>", bucket_name "<bucket_name>", path "<path>"
+    And I create the GCP credentials with a project id "<repl_project_id>" and credentials path "<repl_credentials_path>"
```
Comment (ahmadnazeri) [inline tests/end_to_end/features/provider_updates.feature:72]: any reason why we register it twice?

File: tests/end_to_end/features/steps/filestore_upload.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
+from behave import *
```
Comment (ahmadnazeri) [inline tests/end_to_end/features/steps/filestore_upload.py:1]: for this file, does file stores not support uploading a file? If not, then this is fine. And maybe when/if we support it then we can change this to use that API.

**Review by ahmadnazeri — APPROVED** [review]

File: tests/end_to_end/features/provider_updates.feature (hunk 6 above)
Comment (sdreyer) [reply tests/end_to_end/features/provider_updates.feature:8]: These tests are testing updates, so it recreates the credentials again

File: tests/end_to_end/features/steps/filestore_upload.py (hunk 10 above)
Comment (sdreyer) [reply tests/end_to_end/features/steps/filestore_upload.py:1]: Yeah that was the goal

**Pull request merged**